File: ui/components/free_log.py
"""Free-scrolling execution log element.

NiceGUI's ``ui.log`` snaps the viewport back to the newest line whenever any
content arrives, so an actively streaming execution log cannot be scrolled
up to read earlier lines. ``FreeLog`` swaps in the sibling ``free_log.js``
component: the view follows the tail only while it already rests at the
bottom, and otherwise stays put (line appends happen below it, and trims of
old lines at the ``max_lines`` cap are compensated). The Python API is
inherited unchanged from ``ui.log`` (``push``, ``max_lines``, ``clear``),
so callers only need to swap the class.
"""

from nicegui.elements.log import Log


class FreeLog(Log, component='free_log.js'):
    """Drop-in replacement for ``ui.log`` with a freely scrollable viewport."""

File: ui/components/free_log.js
// Free-scrolling log console (pairs with free_log.py).
//
// NiceGUI's stock log component sets the scroll position to the bottom on
// every update: during streaming output that snap repeatedly wins over the
// user's wheel events, so the log cannot be scrolled up while lines keep
// arriving. Here the view follows the tail only while it already rests at
// the bottom; otherwise it is left alone. Scrolling back near the bottom
// resumes following.
const STICK_TOLERANCE_PX = 32;

export default {
  template: `<div ref="qRef" @scroll="onScroll"><slot></slot></div>`,
  data() {
    return {
      pinnedToBottom: true,
      lastScrollHeight: NaN,
    };
  },
  methods: {
    onScroll() {
      const el = this.$el;
      const distanceFromBottom = el.scrollHeight - el.scrollTop - el.clientHeight;
      this.pinnedToBottom = distanceFromBottom < STICK_TOLERANCE_PX;
    },
  },
  beforeUpdate() {
    // The pre-patch scrollHeight is the baseline for detecting how much
    // content this update added or removed; after the patch that is no
    // longer reconstructable.
    this.lastScrollHeight = this.$el ? this.$el.scrollHeight : NaN;
  },
  updated() {
    const el = this.$el;
    if (this.pinnedToBottom) {
      this.$nextTick(() => {
        el.scrollTop = el.scrollHeight;
      });
      return;
    }
    // Appends land below the viewport and must not move it. Trimming at the
    // max_lines cap removes the oldest lines above the viewport, which would
    // drag the text being read upwards by their height; shift back so the
    // visible text stays stationary.
    this.$nextTick(() => {
      if (!Number.isNaN(this.lastScrollHeight)) {
        const heightDelta = el.scrollHeight - this.lastScrollHeight;
        if (heightDelta < 0) el.scrollTop += heightDelta;
      }
    });
  },
};
